misc-scripts/xref_mapping/XrefParser/RefSeqGPFFParser.pm

```diff
@@ -17,340 +17,330 @@ limitations under the License.
+
+  $self->{name2species_id} = { map{ $_=>$species_id } @{$species2name{$species_id}} };
+  my %species2tax  = $self->species_id2taxonomy($dbi);
+  $self->{taxonomy2species_id} = { map{ $_=>$species_id } @{$species2tax{$species_id}} };
+
+  # process the source files
+  foreach my $file (@{$files}) {
+
+    # type from the file (peptide or dna)
+    my $type = $self->type_from_file($file);
 
-    my @xrefs;
-    foreach my $file (@files) {
-        my $xrefs =
-          $self->create_xrefs( $file, $species_id, $verbose, $dbi, $species_name );
+    # get the file handler
+    my $refseq_fh = $self->get_filehandle($file);
 
-        if ( !defined( $xrefs ) ) {
-            return 1;    #error
-        }
-        $self->upload_xref_object_graphs( $xrefs, $dbi )
+    if ( !defined $refseq_fh ) {
+      warn "WARNING: Can't open RefSeqGPFF file $file\n";
+      return;
     }
 
-    if ( defined $release_file ) {
-        # Parse and set release info.
-        my $release_io = $self->get_filehandle($release_file);
-        local $/ = "\n*";
-        my $release = $release_io->getline();
-        $release_io->close();
-
-        $release =~ s/\s{2,}/ /g;
-        $release =~ s/.*(NCBI Reference Sequence.*) Distribution.*/$1/s;
-        # Put a comma after the release number to make it more readable.
-        $release =~ s/Release (\d+)/Release $1,/;
-
-        print "RefSeq release: '$release'\n" if($verbose);
-
-        $self->set_release( $source_id,              $release, $dbi );
-        $self->set_release( $peptide_source_id,      $release, $dbi );
-        $self->set_release( $mrna_source_id,         $release, $dbi );
-        $self->set_release( $ncrna_source_id,        $release, $dbi );
-        $self->set_release( $pred_mrna_source_id,    $release, $dbi );
-        $self->set_release( $pred_ncrna_source_id,   $release, $dbi );
-        $self->set_release( $pred_peptide_source_id, $release, $dbi );
+    # this will hold the array of xrefs to bulk insert
+    my $xrefs;
+
+    do {
+      local $/ = "\/\/\n";
+      while ( my $genbank_rec = $refseq_fh->getline() ) {
+        my $xref = $self->xref_from_record( $genbank_rec, $type );
+        push @{$xrefs}, $xref if $xref;
+      }
+    };
+
+    $refseq_fh->close();
+
+    # upload the xrefs
+    $self->upload_xref_object_graphs( $xrefs, $dbi ) if $xrefs;
+
+  }
+
+
+  # process the release file
+  if ( defined $release_file ) {
+    # get filehandle
+    my $release_fh = $self->get_filehandle($release_file);
+
+    # get file header
+    my $release = do { local $/ = "\n*"; <$release_fh> };
+    $release_fh->close();
+
+    $release =~ s/\s+/ /xg;
+
+    if ( $release =~ m/(NCBI.*Release\s\d+)\s(.*)\sDistribution/x ) {
+      my ($rel_number, $rel_date) = ($1, $2);
+      my $release_string = "$rel_number, $rel_date";
+
+      # set release info
+      $self->set_release( $source_id, $release_string, $dbi );
+      for my $source_name (sort values %{$REFSEQ_SOURCES}) {
```

Comment:
Does it really matter if we set releases on sources in a specific order? I would say no _i.e._ there is no need to sort the values.